packages/core/src/sync/transport.ts

```diff
@@ -45,7 +60,122 @@ export const cachedTransport = ({
       async request({ method, params }) {
         const body = { method, params };
 
+        // multicall
         if (
+          method === "eth_call" &&
+          params[0]?.data?.startsWith(MULTICALL_SELECTOR)
+        ) {
+          let blockNumber: Hex | "latest" | undefined = undefined;
+          [, blockNumber] = params;
+
+          const multicallData = decodeFunctionData({
+            abi: multicall3Abi,
+            data: params[0]!.data,
+          });
+          const requests = multicallData.args[0]!.map((call) => ({
+            method: "eth_call",
+            params: [
+              {
+                data: call.callData,
+                to: call.target,
+              },
+              blockNumber,
+            ],
+          })).map((request) =>
+            toLowerCase(JSON.stringify(orderObject(request))),
+          );
+
+          if (requests.length === 0) {
+            return encodeFunctionResult({
+              abi: multicall3Abi,
+              functionName: "aggregate3",
+              // @ts-expect-error known issue in viem
+              result: [[]],
+            });
+          }
+
+          const cachedResults = await syncStore.getRpcRequestResults({
+            requests,
+            chainId: chain!.id,
+          });
+
+          const multicallResult = await requestQueue
+            .request({
+              method: "eth_call",
+              params: [
+                {
+                  to: params[0]!.to,
+                  data: encodeFunctionData({
+                    abi: multicall3Abi,
+                    functionName: "aggregate3",
+                    args: [
+                      multicallData.args[0]!.filter(
+                        (_, index) => cachedResults[index] === undefined,
+                      ),
+                    ],
+                  }),
+                },
+                blockNumber!,
+              ],
+            })
+            .then((result) =>
+              decodeFunctionResult({
+                abi: multicall3Abi,
+                functionName: "aggregate3",
+                data: result,
+              }),
+            );
+
+          // Note: insertRpcRequestResults errors can be ignored and not awaited, since
+          // the response is already fetched.
+          syncStore
+            .insertRpcRequestResults({
+              requests: requests
+                .filter((_, index) => cachedResults[index] === undefined)
+                .map((request, index) => ({
+                  request,
+                  result: multicallResult[index]!,
+                }))
+                // Note: we don't cache request that failed or returned "0x". See more about "0x" below.
+                .filter(
+                  ({ result }) => result?.success && result.returnData !== "0x",
+                )
```

Comment:
@typedarray I don't think we should cache failed requests and just re-request them

Reply:
Brilliant